### PR TITLE
Http4s transaction changes

### PR DESCRIPTION
**Pull request by richard-gibson**
_Before contributing, please read our [contributing guidelines](https://github.com/newrelic/newrelic-java-agent/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/newrelic/.github/blob/main/CODE_OF_CONDUCT.md)._

### Overview
This PR adds support for HTTP4s ember and blaze v0.23 client and server libs. It also fixes a bug where transaction activity was not being cleared from the request thread causing incorrect transaction metrics.

### Related Github Issue
Include a link to the related GitHub issue, if applicable

### Testing
The agent includes a suite of tests which should be used to
verify your changes don't break existing functionality. These tests will run with
Github Actions when a pull request is made. More details on running the tests locally can be found
[here](https://github.com/newrelic/newrelic-java-agent/blob/main/CONTRIBUTING.md),

### Checks

[N] Are your contributions backwards compatible with relevant frameworks and APIs? Removal of support transaction support for cats effect parTraverse was necessary to ensure Transaction Activity could be clear from threads after work. 
Cats effect async features still have transaction support.
[N] Does your code contain any breaking changes? Please describe. 
[Y] Does your code introduce any new dependencies? Please describe.
New instrumentation for HTTP4 blaze and ember v0.23 libs

### CODE REVIEW

File: instrumentation/cats-effect-3.3.0/src/main/scala/cats/effect/IOFiber_Instrumentation.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  @NewField
+  private val tokenAndRefCount: AtomicReference[AgentBridge.TokenAndRefCount] =
+    new AtomicReference(getThreadTokenAndRefCount(fiberRef))
+
+  incrementTokenRefCount(this.tokenAndRefCount.get())
+  logTokenInfo(tokenAndRefCount.get(), fiberRef, "IOFiber token info set")
+  private val IOEndFiber: IO.EndFiber.type = Weaver.callOriginal();
+
+  private def fiberRef: IOFiber[A] = this.asInstanceOf[IOFiber[A]]
+  def run(): Unit = {
```
Comment (meiao) [inline instrumentation/cats-effect-3.3.0/src/main/scala/cats/effect/IOFiber_Instrumentation.scala:32]: can you add a line between these functions?
reading thru this and getting confused by which is the function for this body.

File: instrumentation/cats-effect-3.3.0/src/main/scala/cats/effect/IOFiber_Instrumentation.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    val fiberTokenEmpty = this.tokenAndRefCount.get() == null
+    try {
+      if (!fiberTokenEmpty) setThreadTokenAndRefCount(tokenAndRefCount, fiberRef)
+      Weaver.callOriginal
+    } finally {
+      val threadTokenAndRefCount = getThreadTokenAndRefCount(fiberRef)
+      if (fiberTokenEmpty && threadTokenAndRefCount != null) {
+        if (threadTokenAndRefCount != null) {
+          this.tokenAndRefCount.set(threadTokenAndRefCount)
+        }
```
Comment (meiao) [inline instrumentation/cats-effect-3.3.0/src/main/scala/cats/effect/IOFiber_Instrumentation.scala:42]: the 2nd if statement is redundant

File: instrumentation/cats-effect-3.3.0/src/main/scala/cats/effect/IOFiber_Instrumentation.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        incrementTokenRefCount(this.tokenAndRefCount.get())
```
Comment (meiao) [inline instrumentation/cats-effect-3.3.0/src/main/scala/cats/effect/IOFiber_Instrumentation.scala:43]: Is the Weaver.callOriginal() running async and the decrement in done the reverse of this increment?

**Review by meiao — CHANGES_REQUESTED** [review]

File: instrumentation/cats-effect-2/src/main/scala/cats/effect/ec/TokenAwareExecutionContext.scala [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.logging.Level
+import scala.concurrent.ExecutionContext
+
+class TokenAwareExecutionContext(delegate: ExecutionContext) extends ExecutionContext  {
+  AgentBridge.getAgent.getLogger.log(Level.INFO, s"[${Thread.currentThread().getName}] Instrumenting IOShift " +
+    s"ExecutionContext $delegate")
+  override def execute(runnable: Runnable): Unit = {
+    delegate.execute(new TokenAwareRunnable(runnable))
+  }
+  override def reportFailure(cause: Throwable): Unit = delegate.reportFailure(cause)
```
[the remark this answers is not in the document]
Comment (richard-gibson) [reply instrumentation/cats-effect-2/src/main/scala/cats/effect/ec/TokenAwareExecutionContext.scala:14]: Yes, I think so, we don't want to lose any custom reportFailure that could be specified by a user in the delegate Execution context

[the remark this answers is not in the document]
Comment (meiao) [reply instrumentation/cats-effect-2/src/main/scala/cats/effect/ec/TokenAwareExecutionContext.scala:14]: Nevermind, I was really tired at this point and missed that there is a TokenAwareRunnable in there.

File: instrumentation/http4s-ember-client-2.13_0.23/src/main/scala/com/nr/instrumentation/http4s/NewrelicClientMiddleware.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      for {
+        seg <- Resource.eval(
+          construct {
+          val txn = AgentBridge.getAgent.getTransaction
+          logTokenInfo(AgentBridge.activeToken.get, s"client call for txn $txn")
+          val segment = txn.startSegment("HTTP4S client call")
+          segment.addOutboundRequestHeaders(new OutboundRequestWrapper(req))
+          segment
+        })
+        response <- client.run(req)
```
Comment (meiao) [inline instrumentation/http4s-ember-client-2.13_0.23/src/main/scala/com/nr/instrumentation/http4s/NewrelicClientMiddleware.scala:27]: Won't these run async?

File: instrumentation/http4s-ember-server-2.12_0.23/src/main/java/org/http4s/server/blaze/EmberServerBuilder_Instrumentation.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package org.http4s.server.blaze;
```
Comment (meiao) [inline instrumentation/http4s-ember-server-2.12_0.23/src/main/java/org/http4s/server/blaze/EmberServerBuilder_Instrumentation.java:1]: Wrong package

File: settings.gradle [hunk 7]
```diff
@@ -276,7 +286,8 @@ include 'instrumentation:scala-2.9.3'
 include 'instrumentation:scala-2.12.0'
 include 'instrumentation:scala-2.13.0'
 include 'instrumentation:cats-effect-2'
-include 'instrumentation:cats-effect-3'
+include 'instrumentation:cats-effect-3.2'
```
Comment (meiao) [inline settings.gradle:289]: Probably already fixed, but cats-effect-3 is missing.

**Review by meiao — APPROVED** [review]

**Pull request merged**